In [ ]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns 

In [ ]:
purchase_behavior = pd.read_csv(r"/contents/purchase_behaviour.csv")
transaction_data = pd.read_csv(r"/contents/transaction_data.csv")

In [4]:
purchase_behavior.head()

LYLTY_CARD_NBR               LIFESTAGE PREMIUM_CUSTOMER
0            1000   YOUNG SINGLES/COUPLES          Premium
1            1002   YOUNG SINGLES/COUPLES       Mainstream
2            1003          YOUNG FAMILIES           Budget
3            1004   OLDER SINGLES/COUPLES       Mainstream
4            1005  MIDAGE SINGLES/COUPLES       Mainstream

In [5]:
purchase_behavior.shape

(72637, 3)

In [7]:
purchase_behavior.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 72637 entries, 0 to 72636
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   LYLTY_CARD_NBR    72637 non-null  int64 
 1   LIFESTAGE         72637 non-null  object
 2   PREMIUM_CUSTOMER  72637 non-null  object
dtypes: int64(1), object(2)
memory usage: 1.7+ MB


In [11]:
purchase_behavior.duplicated().sum()

np.int64(0)

In [12]:
purchase_behavior['LIFESTAGE'].value_counts()

LIFESTAGE
RETIREES                  14805
OLDER SINGLES/COUPLES     14609
YOUNG SINGLES/COUPLES     14441
OLDER FAMILIES             9780
YOUNG FAMILIES             9178
MIDAGE SINGLES/COUPLES     7275
NEW FAMILIES               2549
Name: count, dtype: int64

In [13]:
purchase_behavior['PREMIUM_CUSTOMER'].value_counts()

PREMIUM_CUSTOMER
Mainstream    29245
Budget        24470
Premium       18922
Name: count, dtype: int64

In [14]:
transaction_data.head()

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0  43390          1            1000       1         5   
1  43599          1            1307     348        66   
2  43605          1            1343     383        61   
3  43329          2            2373     974        69   
4  43330          2            2426    1038       108   

                                  PROD_NAME  PROD_QTY  TOT_SALES  
0    Natural Chip        Compny SeaSalt175g         2        6.0  
1                  CCs Nacho Cheese    175g         3        6.3  
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9  
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0  
4  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8

In [15]:
transaction_data.shape

(264836, 8)

In [16]:
transaction_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 264836 entries, 0 to 264835
Data columns (total 8 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   DATE            264836 non-null  int64  
 1   STORE_NBR       264836 non-null  int64  
 2   LYLTY_CARD_NBR  264836 non-null  int64  
 3   TXN_ID          264836 non-null  int64  
 4   PROD_NBR        264836 non-null  int64  
 5   PROD_NAME       264836 non-null  object 
 6   PROD_QTY        264836 non-null  int64  
 7   TOT_SALES       264836 non-null  float64
dtypes: float64(1), int64(6), object(1)
memory usage: 16.2+ MB


In [17]:
transaction_data['LYLTY_CARD_NBR'].nunique()

72637

In [18]:
transaction_data['PROD_NAME'].nunique()

114

In [19]:
transaction_data['PROD_NBR'].nunique()  

114

In [28]:
transaction_data['PROD_QTY'].value_counts().reset_index().sort_values(by = 'PROD_QTY',ascending=True)

PROD_QTY   count
1         1   27518
0         2  236039
3         3     430
4         4     397
2         5     450
5       200       2

In [29]:
transaction_data['TOT_SALES'].value_counts().reset_index().sort_values(by = 'TOT_SALES',ascending=True)

TOT_SALES  count
39         1.5    354
37         1.7    708
47         1.8    175
34         1.9    930
36         2.1    732
..         ...    ...
75        25.5     14
59        27.0     35
79        28.5     12
91        29.5      7
110      650.0      2

[112 rows x 2 columns]

In [33]:
transaction_data['PROD_NAME'].nunique()

114

In [31]:
data = transaction_data.merge(purchase_behavior,on='LYLTY_CARD_NBR',how='left')

In [32]:
data.head()

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0  43390          1            1000       1         5   
1  43599          1            1307     348        66   
2  43605          1            1343     383        61   
3  43329          2            2373     974        69   
4  43330          2            2426    1038       108   

                                  PROD_NAME  PROD_QTY  TOT_SALES  \
0    Natural Chip        Compny SeaSalt175g         2        6.0   
1                  CCs Nacho Cheese    175g         3        6.3   
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9   
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0   
4  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8   

                LIFESTAGE PREMIUM_CUSTOMER  
0   YOUNG SINGLES/COUPLES          Premium  
1  MIDAGE SINGLES/COUPLES           Budget  
2  MIDAGE SINGLES/COUPLES           Budget  
3  MIDAGE SINGLES/COUPLES           Budget  
4  MIDAGE SINGLES/COUPLES           Budget

In [34]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 264836 entries, 0 to 264835
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   DATE              264836 non-null  int64  
 1   STORE_NBR         264836 non-null  int64  
 2   LYLTY_CARD_NBR    264836 non-null  int64  
 3   TXN_ID            264836 non-null  int64  
 4   PROD_NBR          264836 non-null  int64  
 5   PROD_NAME         264836 non-null  object 
 6   PROD_QTY          264836 non-null  int64  
 7   TOT_SALES         264836 non-null  float64
 8   LIFESTAGE         264836 non-null  object 
 9   PREMIUM_CUSTOMER  264836 non-null  object 
dtypes: float64(1), int64(6), object(3)
memory usage: 20.2+ MB


Most profitable product according Total_Sales

In [40]:
data.groupby('PROD_NAME')['LYLTY_CARD_NBR'].count().reset_index().sort_values(by='PROD_NAME', ascending=False).head(3)


PROD_NAME  LYLTY_CARD_NBR
113  Woolworths Mild     Salsa 300g            1491
112  Woolworths Medium   Salsa 300g            1430
111  Woolworths Cheese   Rings 190g            1516

Most profitable product according to Product Quantity 

In [44]:
data.groupby('PROD_NAME')['PROD_QTY'].sum().reset_index().sort_values(by='PROD_NAME', ascending=False).head(3)

PROD_NAME  PROD_QTY
113  Woolworths Mild     Salsa 300g      2823
112  Woolworths Medium   Salsa 300g      2700
111  Woolworths Cheese   Rings 190g      2872

Customers by Frequency - This shows the customer which are loyal and there characteristics

In [94]:
agg_df = data.groupby(['LIFESTAGE', 'PREMIUM_CUSTOMER'])['LYLTY_CARD_NBR'].value_counts().reset_index(name='count')
top_3 = agg_df.sort_values(by='count', ascending=False).groupby(['LIFESTAGE', 'PREMIUM_CUSTOMER']).head(3)
top_3.head(2)

LIFESTAGE PREMIUM_CUSTOMER  LYLTY_CARD_NBR  count
61975  YOUNG SINGLES/COUPLES       Mainstream          172032     18
14499         OLDER FAMILIES       Mainstream          162039     18

In [90]:
print(top_3['count'].max())

18


In [ ]:
top_3 = agg_df.sort_values(by='count', ascending=False).groupby(['LIFESTAGE', 'PREMIUM_CUSTOMER']).head(3)
most_loyal_customers = top_3.loc[top_3['count'] == top_3['count'].max()]
loyal_customer_data = data[data['LYLTY_CARD_NBR'].isin(most_loyal_customers['LYLTY_CARD_NBR'])]

loyal_customer_summary = loyal_customer_data.groupby(['LIFESTAGE', 'PREMIUM_CUSTOMER']).agg(
    total_sales=('TOT_SALES', 'sum'),
    avg_sales=('TOT_SALES', 'mean'),
    count=('LYLTY_CARD_NBR', 'count')
).reset_index()

print(loyal_customer_summary)


               LIFESTAGE PREMIUM_CUSTOMER  total_sales  avg_sales  count
0         OLDER FAMILIES       Mainstream        126.8   7.044444     18
1  YOUNG SINGLES/COUPLES       Mainstream        120.1   6.672222     18


Most Profitable Customers by Monetary

In [78]:
temp_df = data.groupby('LYLTY_CARD_NBR')['TOT_SALES'].sum().reset_index().sort_values(by='TOT_SALES', ascending=False).drop(columns='LYLTY_CARD_NBR').head(3)
temp_df

TOT_SALES
59694     1300.0
60925      138.6
17292      132.8

In [77]:
data.groupby(['LIFESTAGE', 'PREMIUM_CUSTOMER']).agg(
    total_sales=('TOT_SALES', 'sum'),
    count_people=('LIFESTAGE', 'count'),
    avg_sales=('TOT_SALES', 'mean')
).reset_index().sort_values(by=['avg_sales'], ascending=[False])


LIFESTAGE PREMIUM_CUSTOMER  total_sales  count_people  \
1   MIDAGE SINGLES/COUPLES       Mainstream     90803.85         11874   
19   YOUNG SINGLES/COUPLES       Mainstream    157621.60         20854   
14                RETIREES          Premium     97646.05         13096   
11   OLDER SINGLES/COUPLES          Premium    132263.15         17754   
12                RETIREES           Budget    113147.80         15201   
9    OLDER SINGLES/COUPLES           Budget    136769.80         18407   
8           OLDER FAMILIES          Premium     81958.40         11192   
4             NEW FAMILIES       Mainstream     17013.90          2325   
3             NEW FAMILIES           Budget     21928.45          3005   
15          YOUNG FAMILIES           Budget    139345.85         19122   
10   OLDER SINGLES/COUPLES       Mainstream    133393.80         18318   
6           OLDER FAMILIES           Budget    168363.25         23160   
17          YOUNG FAMILIES          Premium     84025.50         11563   
7           OLDER FAMILIES       Mainstream    103445.55         14244   
13                RETIREES       Mainstream    155677.05         21466   
5             NEW FAMILIES          Premium     11491.10          1589   
16          YOUNG FAMILIES       Mainstream     92788.75         12907   
2   MIDAGE SINGLES/COUPLES          Premium     58432.65          8216   
0   MIDAGE SINGLES/COUPLES           Budget     35514.80          5020   
20   YOUNG SINGLES/COUPLES          Premium     41642.10          6281   
18   YOUNG SINGLES/COUPLES           Budget     61141.60          9242   

    avg_sales  
1    7.647284  
19   7.558339  
14   7.456174  
11   7.449766  
12   7.443445  
9    7.430315  
8    7.322945  
4    7.317806  
3    7.297321  
15   7.287201  
10   7.282116  
6    7.269570  
17   7.266756  
7    7.262395  
13   7.252262  
5    7.231655  
16   7.189025  
2    7.112056  
0    7.074661  
20   6.629852  
18   6.615624

The final results are:
- The most profitable 3 products are Woolworths Mild Salsa 300g, Woolworths Medium Salsa 300g, Woolworths Cheese Rings 190g
- The most profitable 3 customers by total sales are 59694, 60925, 17292.
- The last table in here shows the Customer Segementation by the highest avg_sales to target the specific market segment.